client/src/main/java/org/apache/celeborn/client/write/DataPushQueue.java

```diff
@@ -106,10 +110,17 @@ public ArrayList<PushTask> takePushTasks() throws IOException, InterruptedExcept
               oldCapacity = maxInFlight - pushState.inflightPushes(loc.hostAndPushPort());
               workerCapacity.put(loc.hostAndPushPort(), oldCapacity);
             }
+            workerWaitAttempts.putIfAbsent(loc.hostAndPushPort(), new AtomicInteger(0));
             if (oldCapacity > 0) {
               iterator.remove();
               tasks.add(task);
               workerCapacity.put(loc.hostAndPushPort(), oldCapacity - 1);
+            } else if (workerWaitAttempts.get(loc.hostAndPushPort()).get()
+                >= takeTaskMaxWaitAttempts) {
```

Comment:
If the worker already very slowly, we better take only one task to push  otherwise push task may block other worker push in next turn.